Realm/Tests/ObjectTests.m

```diff
@@ -651,26 +652,117 @@ - (void)testObjectSubclass {
     [realm commitWriteTransaction];
     XCTAssertEqualObjects(NSDate.distantFuture, dateObject.dateCol);
+}
 
+- (void)testDateDistantPast {
+    RLMRealm *realm = [RLMRealm defaultRealm];
+    [realm beginWriteTransaction];
+    DateObject *dateObject = [DateObject createInRealm:realm withValue:@[NSDate.distantPast]];
+    [realm commitWriteTransaction];
+    XCTAssertEqualObjects(NSDate.distantPast, dateObject.dateCol);
+}
+
+- (void)testDate50kYears {
+    NSCalendarUnit units = (NSCalendarUnit)(NSCalendarUnitMonth | NSCalendarUnitYear | NSCalendarUnitDay);
+    NSDateComponents *components = [[NSCalendar currentCalendar] components:units fromDate:NSDate.date];
+    components.calendar = [NSCalendar currentCalendar];
+    components.year += 50000;
+
+    RLMRealm *realm = [RLMRealm defaultRealm];
     [realm beginWriteTransaction];
-    NSDate *date = ({
-        NSCalendarUnit units = (NSCalendarUnit)(NSCalendarUnitMonth | NSCalendarUnitYear | NSCalendarUnitDay);
-        NSDateComponents *components = [[NSCalendar currentCalendar] components:units fromDate:NSDate.date];
-        components.calendar = [NSCalendar currentCalendar];
-        components.year += 50000;
-        components.date;
+    DateObject *dateObject = [DateObject createInRealm:realm withValue:@[components.date]];
+    [realm commitWriteTransaction];
+
+    XCTAssertEqualObjects(components.date, dateObject.dateCol);
+}
+
+static void testDatesInRange(NSTimeInterval from, NSTimeInterval to, void (^check)(NSDate *, NSDate *)) {
+    NSDate *date = [NSDate dateWithTimeIntervalSinceReferenceDate:from];
+
+    RLMRealm *realm = [RLMRealm defaultRealm];
+    [realm beginWriteTransaction];
+    DateObject *dateObject = [DateObject createInRealm:realm withValue:@[date]];
+
+    while (from < to) {
+        check(dateObject.dateCol, date);
+        from = nextafter(from, DBL_MAX);
+        date = [NSDate dateWithTimeIntervalSinceReferenceDate:from];
+        dateObject.dateCol = date;
+    }
+    [realm commitWriteTransaction];
+}
+
+- (void)testExactRepresentationOfDatesAroundNow {
+    NSDate *date = [NSDate date];
+    NSTimeInterval time = date.timeIntervalSinceReferenceDate;
+    testDatesInRange(time - .001, time + .001, ^(NSDate *d1, NSDate *d2) {
+        XCTAssertEqualObjects(d1, d2);
+    });
+}
+
+- (void)testExactRepresentationOfDatesAroundDistantFuture {
+    NSDate *date = [NSDate distantFuture];
+    NSTimeInterval time = date.timeIntervalSinceReferenceDate;
+    testDatesInRange(time - .001, time + .001, ^(NSDate *d1, NSDate *d2) {
+        XCTAssertEqualObjects(d1, d2);
+    });
+}
+
+- (void)testExactRepresentationOfDatesAroundEpoch {
+    NSDate *date = [NSDate dateWithTimeIntervalSince1970:0];
+    NSTimeInterval time = date.timeIntervalSinceReferenceDate;
+    testDatesInRange(time - .001, time + .001, ^(NSDate *d1, NSDate *d2) {
+        XCTAssertEqualObjects(d1, d2);
     });
-    dateObject.dateCol = date;
+}
```

Comment:
Would it be worth testing the exact representation of dates around `NSDate`'s epoch as well?